File: src/main/java/com/banking/service/AccountService.java
package com.banking.service;

import com.banking.dto.AccountDTO;

import java.util.List;

import com.banking.security.AuthTokenService.AuthenticatedUser;

/**
 * Service interface for Account operations.
 * Defines the contract for account-related business logic.
 */
public interface AccountService {

    /**
     * Create a new account for a customer.
     */
    AccountDTO createAccount(AccountDTO accountDTO, AuthenticatedUser user);

    /**
     * Get an account by its ID.
     */
    AccountDTO getAccountById(Long id, AuthenticatedUser user);

    /**
     * Get all accounts.
     */
    List<AccountDTO> getAllAccounts();

    /**
     * Get all accounts belonging to a specific customer.
     */
    List<AccountDTO> getAccountsByCustomerId(Long customerId);

    /**
     * Close (delete) an account by its ID.
     */
    void deleteAccount(Long id, AuthenticatedUser user);

    /**
     * Submit a request to delete an account (user action).
     */
    AccountDTO requestAccountDeletion(Long id, AuthenticatedUser user);

    /**
     * Approve and execute account deletion (admin action).
     */
    AccountDTO approveAccountDeletion(Long id, AuthenticatedUser user);

    /**
     * Reject account deletion request (admin action).
     */
    AccountDTO rejectAccountDeletion(Long id, AuthenticatedUser user);

    /**
     * Switch account type between SAVINGS and CURRENT.
     */
    AccountDTO switchAccountType(Long id, AuthenticatedUser user);
}

File: frontend/src/app/services/account.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { environment } from '../../environments/environment';
import { Account } from '../models/account.model';

@Injectable({
  providedIn: 'root'
})
export class AccountService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${environment.apiUrl}/api/accounts`;

  getAccountsByCustomerId(customerId: number): Observable<Account[]> {
    const params = new HttpParams().set('customerId', String(customerId));
    return this.http.get<Account[]>(this.apiUrl, { params });
  }

  getAllAccounts(): Observable<Account[]> {
    return this.http.get<Account[]>(this.apiUrl);
  }

  createAccount(account: Account): Observable<Account> {
    return this.http.post<Account>(this.apiUrl, account);
  }

  deleteAccount(id: number): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }

  requestAccountDeletion(id: number): Observable<Account> {
    return this.http.post<Account>(`${this.apiUrl}/${id}/request-deletion`, {});
  }

  approveAccountDeletion(id: number): Observable<Account> {
    return this.http.post<Account>(`${this.apiUrl}/${id}/approve-deletion`, {});
  }

  rejectAccountDeletion(id: number): Observable<Account> {
    return this.http.post<Account>(`${this.apiUrl}/${id}/reject-deletion`, {});
  }

  switchAccountType(id: number): Observable<Account> {
    return this.http.put<Account>(`${this.apiUrl}/${id}/switch-type`, {});
  }
}
